keps/sig-apps/3017-pod-healthy-policy-for-pdb/README.md

```diff
@@ -0,0 +1,427 @@
+- Make sure users who rely on PDBs for data-safety can continue to do so.
+
+### Non-Goals
+
+- Providing a safe solution for preventing data-loss. Not because this isn't important, but 
+it is unclear if PDB is the right tool for this.
+
+## Proposal
+
+The core issue here is whether a pod that is Running but not Ready is already disrupted, 
+and thus can be evicted without being constrained by a potential Pod Disruption Budget. Currently,
+the disruption controller only considers Running and Ready pods as healthy, thus that is 
+the basis for computing `allowedDisruptions`. The disruption API on the other hand, does require
+that `allowedDisruptions` is larger than 0 to evict a pod that is Running but not Ready. 
+
+Adding a `podReadyPolicy` field on the PDB API will allow the user to specify which
+behavior that are desired, and this will be consistently handled by the disruption controller, eviction
+API, and any other APIs that might use PDBs.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+```golang
+// PodDisruptionBudgetSpec is a description of a PodDisruptionBudget.
+type PodDisruptionBudgetSpec struct {
+	
+	...
+	
+	// PodHealthyPolicy defines the criteria for when the disruption controller
+	// should consider a pod to be healthy.
+	// If no policy is specified, the legacy behavior will be used. It means
+	// only pods that are Running and Ready will be considered when the disruption
+	// controller computes "disruptionsAllowed", but all pods in the Running phase
+	// will be subject to the PDB on eviction.
+	// +optional
+	PodHealthyPolicy PodHealthyPolicy `json:"podHealthyPolicy,omitempty" protobuf:"bytes,4,opt,name=podHealthyPolicy"`
+}
+
+// PodHealthyPolicy defines the policy when a pod are considered healthy and therefore
+// covered by a PodDisruptionBudget.
+type PodHealthyPolicy string
+
+const (
+	// PodReady policy means that only pods that are both Running and Ready
+	// will be considered healthy by the disruption controller. Any pods that
+	// are not Ready are considered to already be disrupted and therefore will
+	// not be counted when computing "disruptionsAllowed" and can be evicted
+	// regardless of whether the criteria in a PDB is met.
+	PodReady PodHealthyPolicy = "PodReady"
+	
+	// PodRunning policy means that pods that are in the Running phase
+	// is considered healthy by the disruption controller, regardless of
+	// whether they are Ready or not. Any pods that are in the Running
+	// phase will be counted when computing "disruptionsAllowed" and
+	// will be subject to the PDB for eviction.
+	PodRunning PodHealthyPolicy = "PodRunning"
+)
+```
+
+### Changes to the disruption controller
+
+The disruption controller will be updated to use `PodHealthyPolicy` to
+determine how it should compute the value of `DisruptionsAllowed`.
+
+### Changes to the eviction API
+
+The eviction API will be updated to use `PodHealthyPolicy` to determine whether
```

Comment:
Would eviction API change as well or just it will be an eviction implementation change.

Reply:
It is just an implementation change that will only matter if users have specified a `podHealthyPolicy`. If they have not, there is not behavior change. I have updated the language here to make it clearer.